### PR TITLE
Add SpannerHealthIndicator and AutoConfiguration

### LINKED ISSUE #640: Health Indicator for Cloud Spanner
Copied from https://github.com/spring-cloud/spring-cloud-gcp/issues/2680 (old repository) by @ikeyat

---

I think Spring Boot's DataSourceHealthIndicator does not work for spring-cloud-gcp-data-spanner because it does not use JDBC DataSource.
A custom health indicator Such as SpannerHealthIndicator and its auto configures should be added into spring-cloud-gcp-data-spanner to avoid boilerplate.

https://docs.spring.io/spring-boot/docs/current/reference/html/actuator.html#actuator.endpoints.health.writing-custom-health-indicators

### PR DESCRIPTION
**Pull request by ikeyat**
Fixes #640

Following tasks have been completed. Please review.

~~Don't merge. Work in progress:~~
 - [x] Test classes are not implemented
 - [x] The documentation is not updated

### CODE REVIEW

**Comment by ikeyat** [comment]
Fix and add tests. Please review.

**Comment by mpeddada1** [comment]
Hey @ikeyat, we still need a few more days to review your PR. Thank you for your patience! We appreciate your contribution.

File: spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+		SpannerHealthIndicator.doHealthCheck(builder);
+
+		assertThat(builder.build().getStatus()).isSameAs(Status.UP);
+		verify(spannerTemplate).executeQuery(Statement.of(QUERY), null);
+		verify(resultSet).next();
+	}
+
+	@Test(expected = Exception.class)
+	public void testdoHealthCheckDownSpannerTemplate() throws Exception {
+		SpannerHealthIndicator SpannerHealthIndicator = new SpannerHealthIndicator(spannerTemplate, QUERY);
```
Comment (elefeint) [inline spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java:70]:
```suggestion
		SpannerHealthIndicator spannerHealthIndicator = new SpannerHealthIndicator(spannerTemplate, QUERY);
```

File: spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+		when(spannerTemplate.executeQuery(any(), any())).thenThrow(new RuntimeException("Cloud Spanner is down!!!"));
+
+		Health.Builder builder = new Health.Builder();
+
+		SpannerHealthIndicator.doHealthCheck(builder);
```
Comment (elefeint) [inline spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java:76]:
```suggestion
		spannerHealthIndicator.doHealthCheck(builder);
```

File: spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	@Test(expected = Exception.class)
+	public void testdoHealthCheckDownResultSet() throws Exception {
+		SpannerHealthIndicator SpannerHealthIndicator = new SpannerHealthIndicator(spannerTemplate, QUERY);
```
Comment (elefeint) [inline spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java:81]:
```suggestion
		spannerHealthIndicator SpannerHealthIndicator = new SpannerHealthIndicator(spannerTemplate, QUERY);
```

File: spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+		when(spannerTemplate.executeQuery(any(), any())).thenReturn(resultSet);
+		when(resultSet.next()).thenThrow(new RuntimeException("Cloud Spanner is down!!!"));
+
+		Health.Builder builder = new Health.Builder();
+
+		SpannerHealthIndicator.doHealthCheck(builder);
```
Comment (elefeint) [inline spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java:88]:
```suggestion
		spannerHealthIndicator.doHealthCheck(builder);
```

File: spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	@Test
+	public void testHealthy() {
+		SpannerHealthIndicator SpannerHealthIndicator = new SpannerHealthIndicator(spannerTemplate, QUERY);
```
Comment (elefeint) [inline spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java:93]:
```suggestion
		spannerHealthIndicator SpannerHealthIndicator = new SpannerHealthIndicator(spannerTemplate, QUERY);
```

File: spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+		when(resultSet.next()).thenReturn(true);
+
+		assertThat(SpannerHealthIndicator.health().getStatus()).isSameAs(Status.UP);
+		verify(spannerTemplate).executeQuery(Statement.of(QUERY), null);
+		verify(resultSet).next();
+	}
+
+	@Test
+	public void testUnhealthySpannerTemplate() {
+		SpannerHealthIndicator SpannerHealthIndicator = new SpannerHealthIndicator(spannerTemplate, QUERY);
```
Comment (elefeint) [inline spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java:105]:
```suggestion
		SpannerHealthIndicator spannerHealthIndicator = new SpannerHealthIndicator(spannerTemplate, QUERY);
```

File: spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+		when(spannerTemplate.executeQuery(any(), any())).thenThrow(new RuntimeException("Cloud Spanner is down!!!"));
+
+		assertThat(SpannerHealthIndicator.health().getStatus()).isEqualTo(Status.DOWN);
```
Comment (elefeint) [inline spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java:109]:
```suggestion
		assertThat(spannerHealthIndicator.health().getStatus()).isEqualTo(Status.DOWN);
```

File: spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		verify(spannerTemplate).executeQuery(Statement.of(QUERY), null);
+		verify(resultSet, never()).next();
+	}
+
+	@Test
+	public void testUnhealthyResultSet() {
+		SpannerHealthIndicator SpannerHealthIndicator = new SpannerHealthIndicator(spannerTemplate, QUERY);
```
Comment (elefeint) [inline spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java:116]:
```suggestion
		SpannerHealthIndicator spannerHealthIndicator = new SpannerHealthIndicator(spannerTemplate, QUERY);
```

File: spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+		when(spannerTemplate.executeQuery(any(), any())).thenReturn(resultSet);
+		when(resultSet.next()).thenThrow(new RuntimeException("Cloud Spanner is down!!!"));
+
+		assertThat(SpannerHealthIndicator.health().getStatus()).isEqualTo(Status.DOWN);
```
Comment (elefeint) [inline spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/spanner/health/SpannerHealthIndicatorTests.java:121]:
```suggestion
		assertThat(spannerHealthIndicator.health().getStatus()).isEqualTo(Status.DOWN);
```

**Review by elefeint — CHANGES_REQUESTED** [review]
Thank you, and sorry for the delay. 
Please add `@since 2.0.6` tags to new classes, and there is one more inline comment. Otherwise looks very good.

**Comment by ikeyat** [comment]
@elefeint 
Thanks for your review. I fixed them.

**Review by elefeint — APPROVED** [review]
This looks great, thank you! Could you add documentation to `spanner.adoc`? Especially around turning the indicator off.

**Comment by ikeyat** [comment]
@elefeint 
I've updated the adoc!

**Comment by elefeint** [comment]
@ikeyat Thank you!

**Pull request merged**